### ISSUE TITLE
test/ddd: sigsegv failure when SSA-built

### ISSUE THREAD

**Issue by binarycrusader** [4ad/go#20]
test/ddd.go currently seg faults when SSA-built and then executed on a sparc64 system.

It is cross-compiled and cross-linked for sparc64 from an amd64 system using:

```bash
SSATEST=1 GOSSAHASH="" GOOS=solaris GOARCH=sparc64 go tool compile ddd.go
GOOS=solaris GOARCH=sparc64 go tool link ddd.o
```

When executed it fails as follows:
```
$ ./ddd
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x8]

goroutine 1 [running]:
panic(0xcdca0, 0x110dc0)
        /builds/srwalker/golang/go-sparc/src/runtime/panic.go:507 +0x5f8
```

**Comment by binarycrusader** [comment]
This appears to be related to interface calls; a reduced example from ddd.go is as follows:

```go
package main

//go:noinline
func sum(args ...int) int {
        s := 0
        for _, v := range args {
                s += v
        }
        return s
}

type T []T 

//go:noinline
func (*T) Sum(args ...int) int { return sum(args...) }

type U struct {
        *T
}

type I interface {
        Sum(...int) int 
}

func main() {
        var u U 
        var i interface {
                Sum(...int) int 
        } = &u
        if x := i.Sum(2, 3, 5, 7); x != 17 {
                panic("fail")
        }   
}
```

**Comment by binarycrusader** [comment]
progedit's mangling of operations that reference %sp appears to be causing code generation problems.  If I build the ddd test so that the only function that is SSA-built is the Interface Sum function, the test will fail.

```GOSSAFUNC='(*U).Sum' SSATEST=1 GOSSAHASH="" GOOS=solaris GOARCH=sparc64 go tool compile ddd.go```

Comparing the old and new disassembly, the only meaningful differences are:
1. new code setups a stack frame 176 bytes in size; old code generates a leaf function (no stack frame)
1. new code stores what are presumably arguments to the sum() function at the offsets of 0x95f, 0x97f; old code stores them at the expected offsets of 0x8af, 0x8cf

This may be misleading though and only wrong because I'm selectively applying the SSA-based backend to only the interface Sum() function.

**Comment by binarycrusader** [comment]
The problem seems to be that `main.(*U).Sum` now has a stack frame, but `main.(*T).Sum` fails to account for that stack frame when attempting to reference items on main.main's stack.  As such, the addresses generated are wrong; they're offset by exactly 0xb0 from where they should be.

**Comment by binarycrusader** [comment]
I've been working with a further simplified example:

```go
package main

//go:noinline
func sum(args ...int) int {
        l := len(args)
        if l != 1 { 
                println("len(args) should be 1; is", l)
        }
        return 1
}

type T []T 

func (*T) Sum(args ...int) int {
        return sum(args...)
}

type U struct {
        *T
}

type I interface {
        Sum(...int) int 
}

var u U 
var i interface {
        Sum(...int) int 
} = &u

func main() {
        if x := i.Sum(1); x != 1 { 
                panic("fail")
        }
}
````

Typically, this fails with something like: `len(args) should be 1; is 72339911365402048` which is unsurprising since the wrong part of the stack is being referenced.

**Comment by binarycrusader** [comment]
I don't know whether to whether to pursue determining why `main.(*U).Sum` now has a stack frame and fix that, or to figure out why the wrong stack address is generated for `main.(*T).Sum` -- both seem like things that should be fixed.  It's likely that fixing the latter will be simpler and inform the former.

**Comment by binarycrusader** [comment]
This was because we were using AJMP instead of ARET in the ssa backend when generating the code for a BlockRetJmp.  The obj backend for sparc64 has a special case for a return with a symbol; this also ensures the function gets marked as a leaf later on.

Fixing this resolves both #20 and #24.
